shell/platform/android/io/flutter/embedding/android/FlutterView.java

```diff
@@ -520,18 +530,47 @@ public final WindowInsets onApplyWindowInsets(@NonNull WindowInsets insets) {
       if (statusBarVisible) {
         mask = mask | android.view.WindowInsets.Type.statusBars();
       }
-      mask = mask | android.view.WindowInsets.Type.ime();
-
-      Insets finalInsets = insets.getInsets(mask);
-      viewportMetrics.paddingTop = finalInsets.top;
-      viewportMetrics.paddingRight = finalInsets.right;
-      viewportMetrics.paddingBottom = 0;
-      viewportMetrics.paddingLeft = finalInsets.left;
+      Insets uiInsets = insets.getInsets(mask);
+      viewportMetrics.paddingTop = uiInsets.top;
+      viewportMetrics.paddingRight = uiInsets.right;
+      viewportMetrics.paddingBottom = uiInsets.bottom;
+      viewportMetrics.paddingLeft = uiInsets.left;
+
+      Insets imeInsets = insets.getInsets(android.view.WindowInsets.Type.ime());
+      viewportMetrics.viewInsetTop = imeInsets.top;
+      viewportMetrics.viewInsetRight = imeInsets.right;
+      viewportMetrics.viewInsetBottom = imeInsets.bottom; // Typically, only bottom is non-zero
+      viewportMetrics.viewInsetLeft = imeInsets.left;
+
+      Insets systemGestureInsets = insets.getInsets(android.view.WindowInsets.Type.systemGestures());
+      viewportMetrics.systemGestureInsetTop = systemGestureInsets.top;
+      viewportMetrics.systemGestureInsetRight = systemGestureInsets.right;
+      viewportMetrics.systemGestureInsetBottom = systemGestureInsets.bottom;
+      viewportMetrics.systemGestureInsetLeft = systemGestureInsets.left;
 
-      viewportMetrics.viewInsetTop = 0;
-      viewportMetrics.viewInsetRight = 0;
-      viewportMetrics.viewInsetBottom = finalInsets.bottom;
-      viewportMetrics.viewInsetLeft = 0;
+      // TODO(garyq): Expose the full rects of the display cutout.
+
+      // Take the max of the display cutout insets and existing padding to merge them
+      DisplayCutout cutout = insets.getDisplayCutout();
```

Comment:
Oh, neat that they have the insets apis separate from the rect apis. I suppose there's no regression here since on iOS, you already can't tell apart the notch rect from the insets to "utilize" the remaining area. 

Let's do the rect separately, sure.

Reply:
https://github.com/flutter/flutter/issues/65088 Tracks the rect work